pkg/dynamicinstrumentation/diconfig/mem_config_manager.go

```diff
@@ -0,0 +1,216 @@
+type configsByService = map[ditypes.ServiceName]map[ditypes.ProbeID]rcConfig
+
+func NewReaderConfigManager() (*ReaderConfigManager, error) {
+	cm := &ReaderConfigManager{
+		callback: applyConfigUpdate,
+		state:    ditypes.NewDIProcs(),
+	}
+
+	cm.procTracker = proctracker.NewProcessTracker(cm.updateProcessInfo)
+	err := cm.procTracker.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	reader := NewConfigReader(cm.updateServiceConfigs)
+	err = reader.Start()
+	if err != nil {
+		return nil, err
+	}
+	cm.ConfigReader = reader
+	return cm, nil
+}
+
+func (cm *ReaderConfigManager) GetProcInfos() ditypes.DIProcs {
+	return cm.state
+}
+
+func (cm *ReaderConfigManager) Stop() {
+	cm.ConfigReader.Stop()
+	cm.procTracker.Stop()
+}
+
+func (cm *ReaderConfigManager) update() error {
+	var updatedState = ditypes.NewDIProcs()
+	for serviceName, configsByID := range cm.configs {
+		for pid, proc := range cm.ConfigReader.Processes {
+			// If a config exists relevant to this proc
+			if proc.ServiceName == serviceName {
+				procCopy := *proc
+				updatedState[pid] = &procCopy
+				updatedState[pid].ProbesByID = convert(serviceName, configsByID)
+			}
+		}
+	}
+
+	if !reflect.DeepEqual(cm.state, updatedState) {
+		err := inspectGoBinaries(updatedState)
+		if err != nil {
+			return err
+		}
+
+		for pid, procInfo := range cm.state {
+			// cleanup dead procs
+			if _, running := updatedState[pid]; !running {
+				procInfo.CloseAllUprobeLinks()
+				delete(cm.state, pid)
+			}
+		}
+
+		for pid, procInfo := range updatedState {
+			if _, tracked := cm.state[pid]; !tracked {
+				for _, probe := range procInfo.GetProbes() {
+					// install all probes from new process
+					cm.callback(procInfo, probe)
+				}
+			} else {
+				currentStateProbes := cm.state[pid].GetProbes()
+				for _, existingProbe := range currentStateProbes {
+					cm.state[pid].DeleteProbe(existingProbe.ID)
+				}
+				for _, updatedProbe := range procInfo.GetProbes() {
+					cm.callback(procInfo, updatedProbe)
+				}
+			}
+		}
+		cm.state = updatedState
+	}
+	return nil
+}
+
+func (cm *ReaderConfigManager) updateProcessInfo(procs ditypes.DIProcs) {
+	cm.Lock()
+	defer cm.Unlock()
+	log.Info("Updating procs", procs)
+	cm.ConfigReader.UpdateProcesses(procs)
+	err := cm.update()
+	if err != nil {
+		log.Info(err)
+	}
+}
+
+func (cm *ReaderConfigManager) updateServiceConfigs(configs configsByService) {
+	cm.configs = configs
+	err := cm.update()
+	if err != nil {
+		log.Info(err)
+	}
+}
+
+type ConfigReader struct {
+	io.Reader
+	updateChannel  chan ([]byte)
+	Processes      map[ditypes.PID]*ditypes.ProcessInfo
+	configCallback configReaderCallback
+	stopChannel    chan (bool)
+}
+
+type configReaderCallback func(configsByService)
+
+func NewConfigReader(onConfigUpdate configReaderCallback) *ConfigReader {
+	return &ConfigReader{
+		updateChannel:  make(chan []byte),
+		configCallback: onConfigUpdate,
+	}
+}
+
+func (r *ConfigReader) Read(p []byte) (n int, e error) {
```

Comment:
This is a confusing implementation of `io.Reader`. Normally data is read and written into `p`, whereas here you are taking data from `p` and sending it elsewhere. I think this should be an `io.Writer` instead.

Reply:
Makes sense